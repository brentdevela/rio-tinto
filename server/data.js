const ideas = [{
    id: '1',
    createdDate: '2020-02-01',
    title: 'My Tile',
    body: 'This is my idea!'
},{
    id: '2',
    createdDate: '2020-02-02',
    title: 'Your Tile',
    body: 'This was your idea!'
},{
    id: '3',
    createdDate: '2020-02-03',
    title: 'Their Tile',
    body: 'This was their idea!'
},{
    id: '4',
    createdDate: '2020-02-04',
    title: 'Silly Tile',
    body: 'What a silly idea!'
},{
    id: '5',
    createdDate: '2020-02-05',
    title: 'Boring Tile',
    body: 'No idea!'
},{
    id: '6',
    createdDate: '2020-02-06',
    title: 'Happy Tile',
    body: 'Great idea!!'
},{
    id: '7',
    createdDate: '2020-02-07',
    title: 'Sad Tile',
    body: 'That was a conservative idea!'
},{
    id: '8',
    createdDate: '2020-02-08',
    title: 'Angry Tile',
    body: 'What a liberal idea!'
},{
    id: '9',
    createdDate: '2020-02-09',
    title: 'Progressive Tile',
    body: 'Isnt that a progressive idea!'
},{
    id: '10',
    createdDate: '2020-02-10',
    title: 'Crazy Tile',
    body: 'What a liberal idea!'
},{
    id: '11',
    createdDate: '2020-02-11',
    title: 'Indifferent Tile',
    body: 'What an indifferent idea!'
}];

const newIdea = {
    id: '',
    createdDate: '',
    title: '',
    body: ''
};

module.exports = {
    ideas,
    newIdea
};

