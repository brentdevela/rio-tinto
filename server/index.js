'use strict';
const port = 8010;
const server = require('./server')();
server.listen(port, () => console.log(`App started and listening on port ${port}`));