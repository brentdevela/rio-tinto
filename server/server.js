'use strict';
const lodash = require('lodash');
const moment = require('moment');
const express = require('express');
const cors = require('cors');
const app = express();

app.use(cors());

const bodyParser = require('body-parser');
const jsonParser = bodyParser.json();

const data = require('./data');

module.exports = () => {
    app.get('/ideas', (req, res) => {
        res.send(data.ideas); // DUMMY
    });

    app.post('/idea', jsonParser, (req, res) => {
        const clonedIdea = lodash.cloneDeep(data.newIdea);
        clonedIdea.id = String(Number(lodash.last(data.ideas).id) + 1);
        clonedIdea.createdDate = moment().format('YYYY-MM-DD');
        data.ideas.push(clonedIdea);
        res.send(clonedIdea);
    });

    app.put('/idea', jsonParser, (req, res) => {
        const idea = req.body;
        const index = lodash.findIndex(data.ideas, (currentIdea) => currentIdea.id === idea.id);
        data.ideas[index] = idea;
        res.send(req.body);
    });

    app.del('/idea/:id', jsonParser, (req, res) => {
        lodash.remove(data.ideas, (idea) => idea.id === req.params.id);
        res.send(req.body);
    });

    return app;
};
