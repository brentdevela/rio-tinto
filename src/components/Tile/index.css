.tile {
    width: 150px;
    height: 150px;
    margin: 10px;
    padding: 2px 5px 5px 5px;
    border: 1px solid darkgray;
    text-align: left;
}

.tile .mb-3 {
    margin-bottom: 0.15rem !important;
}

.tile .form-control {
    border: 0px solid #ced4da;
}

.tile .form-control:focus {
    border: 1px solid lightgrey;
    box-shadow: none;
}

.tile:hover .close {
    color: black;
}

.tile textarea {
    resize: none;
}

.header {
    display: flex;
    font-size: 12px;
}

.header .id {
    width: 95%;
    padding-top: 8px;
}

.header .close {
    color: white;
}

.footer {
    font-size: 10px;
    margin-bottom: 5px;
    text-align: right;
}

