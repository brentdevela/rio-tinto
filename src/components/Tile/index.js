import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { InputGroup, FormControl } from 'react-bootstrap';
import * as ReactDOM from 'react-dom';
import { cloneDeep, noop } from 'lodash';
import { IoIosClose } from 'react-icons/io';
import Idea from '../../models/Idea';

import './index.css';

const Tile = (props) => {

    const [ idea, setIdea ] = useState(props.idea);
    let title;
    useEffect(() => {
        if (props.isFocused) {
            ReactDOM.findDOMNode(title).focus();
        }
    });

    const onClickDelete = () => {
      props.onDeleteIdea(props.idea);
    };

    const onInputChange = (e) => {
        const updatedIdea = cloneDeep(idea);
        updatedIdea[e.target.id] = e.target.value;
        setIdea(updatedIdea);
    };

    const onInputBlur = () => {
      props.onChangeIdea(idea);
    };

    return (
        <div className="tile">
            <div className="header">
                <div className="id">{idea.id}</div>
                <div className="close" onClick={onClickDelete}>
                    <IoIosClose/>
                </div>
            </div>
            <InputGroup size="sm" className="mb-3">
                <FormControl
                    id="title"
                    placeholder="Title"
                    aria-label="Title"
                    type="text"
                    defaultValue={idea.title}
                    onChange={onInputChange}
                    onBlur={onInputBlur}
                    ref={ref => { title = ref; }}
                    maxLength={140}
                />
            </InputGroup>
            <InputGroup size="sm">
                <FormControl
                    id="body"
                    placeholder="Body"
                    aria-label="Body"
                    type="text"
                    as="textarea"
                    rows="3"
                    defaultValue={idea.body}
                    onChange={onInputChange}
                    onBlur={onInputBlur}
                    maxLength={140}
                />
            </InputGroup>
            <div className="footer">
                <div>{idea.createdDate}</div>
            </div>
        </div>
    );
};

Tile.propTypes = {
    idea: Idea.isRequired,
    onChangeIdea: PropTypes.func,
    onDeleteIdea: PropTypes.func,
    isFocused: PropTypes.bool,
};

Tile.defaultProps = {
    onChangeIdea: noop,
    onDeleteIdea: noop,
    isFocused: false,
};

export default Tile;
