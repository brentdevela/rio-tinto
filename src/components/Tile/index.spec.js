import React from 'react';
import { shallow } from 'enzyme';
import { noop } from 'lodash';
import Tile from './index';
import { ideas } from "../../../server/data";

describe('<Tile>', () => {
    const render = (props) => {
        return shallow(
            <Tile
                {...props}
            />
        );
    };

    it('should render Tile', () => {
        const tile = render({ idea:ideas[0] });
        expect(tile).toMatchSnapshot();
    });

    it('should trigger onClickDelete', () => {
        const onDeleteIdea = jest.fn();
        const tile = render({
            idea:ideas[0],
            onDeleteIdea
        });
        const closeButton = tile.find('.close').first();
        closeButton.simulate('click');
        expect(onDeleteIdea).toBeCalled();
    });

    it('should not trigger any prop funcs ', () => {
        const onChangeIdea = jest.fn();
        const tile = render({
            idea:ideas[0],
            onChangeIdea
        });
        const titleField = tile.find('FormControl').first();
        titleField.simulate('change', { target: { id: 'title', value: 'New Title'} });
        expect(onChangeIdea).not.toBeCalled();
    });

    it('should trigger onInputBlur  ', () => {
        const onChangeIdea = jest.fn();
        const tile = render({
            idea:ideas[0],
            onChangeIdea
        });
        const titleField = tile.find('FormControl').first();
        titleField.simulate('change', { target: { id: 'title', value: 'New Title'} });
        titleField.simulate('blur');
        expect(onChangeIdea).toBeCalled();
    });
});