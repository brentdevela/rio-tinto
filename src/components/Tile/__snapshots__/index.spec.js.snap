// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tile> should render Tile 1`] = `ShallowWrapper {}`;
